presto-main-base/src/main/java/com/facebook/presto/sql/rewrite/NativeExecutionTypeRewrite.java

```diff
@@ -183,5 +232,56 @@ private Expression convertEnumTypeToMapExpression(Type type)
                             new ArrayConstructor(keys.build()),
                             new ArrayConstructor(values.build())));
         }
+
+        private Expression convertEnumTypeToLiteral(DereferenceExpression key, Type type)
+        {
+            String enumKey = key.getField().getValue().toUpperCase();
+            if (type instanceof BigintEnumType) {
+                Map<String, Long> enumMap = ((EnumType) type).getEnumMap();
+                Long enumValue = enumMap.get(enumKey);
+                if (enumValue == null) {
+                    throw new SemanticException(TYPE_MISMATCH, "No value " + enumKey + " in enum BigintEnum");
+                }
+                return new LongLiteral(enumValue.toString());
+            }
+            else if (type instanceof VarcharEnumType) {
+                Map<String, String> enumMap = ((EnumType) type).getEnumMap();
+                String enumValue = enumMap.get(enumKey);
+                if (enumValue == null) {
+                    throw new SemanticException(TYPE_MISMATCH, "No value " + enumKey + " in enum VarcharEnum");
+                }
+                return new StringLiteral(enumValue);
+            }
+            return key;
+        }
+
+        private Expression rewriteIfCastOrDereferenceExpression(Expression argument)
+        {
+            if (argument instanceof Cast) {
+                argument = (Expression) visitCast((Cast) argument, null);
+            }
+            return peelIfDereferenceExpression(argument);
+        }
+
+        private Expression peelIfDereferenceExpression(Expression argument)
+        {
+            if (argument instanceof DereferenceExpression) {
+                try {
+                    DereferenceExpression arg = (DereferenceExpression) argument;
+                    Type argumentType = functionAndTypeResolver.getType(parseTypeSignature(arg.getBase().toString()));
+                    if (argumentType instanceof TypeWithName) {
+                        argumentType = ((TypeWithName) argumentType).getType();
+                        if (argumentType instanceof EnumType) {
+                            return convertEnumTypeToLiteral(arg, argumentType);
+                        }
+                    }
+                }
+                catch (IllegalArgumentException | UnknownTypeException e) {
```

Comment:
in what cases does the rewrite fail.  Could we write the code so it doesn't happen instead of relying on exception handling, which could mask bugs.

Reply:
I found that it can fail while getting the type (functionAndTypeResolver.getType) for other dereference expressions, for example for a named field in a row or getting some values from a CTE, as it is not recognized as a valid type. I'm not sure what else we can do if this throws other than return the original node. Would logging help?